README.md
# Gandi Baobab

The application that powers Gandi's Status website ([status.gandi.net](https://status.gandi.net)).

## OVERVIEW

Baobab is a Python and web application divided into 3 parts:
 - a Django-powered back-end (`./baobab/backoffice` and `./baobab/translate`)
 - a REST API (`./baobab/apirest`)
 - a web client that uses the REST API (`./baobab.front`)

It has 5 app namespaces that allow you to work (test, migrate, etc.) on specific parts: `backoffice`, `apirest`, `rss`, `translate` and `socialnetwork`.

Baobab requires a database and supports SQLite, PostgreSQL or MySQL. 

Node.js and npm are required to build the web client app.

Cron jobs are used to publish Tweets and close events on a schedule.

## DEVELOPMENT

### INSTALLATION

Clone the repository and enter the local directory that was just created.

We recommend using [virtualenv](https://virtualenv.pypa.io/en/latest/) to manage local python environments.

    $ virtualenv /some/directory/virtual
    $ source /some/directory/virtual/bin/activate
    $ python setup.py install

A `$ baobab` command should be available after install. Otherwise, please try to use it from the installation directory:
  
    $ cd /some/directory/virtual/lib/python2.7/dist-packages/baobab/bin
    $ ./cmd_baobab.py <...>
  
This can be caused by an error in `django-tastypie`'s `setup.py` distributed in the Debian package.

**Note for Gnome users**: A "baobab" command can already exist in your system (Disk Usage Analyzer), so make sure you use virtualenv as recommended.

### APPLICATION SETTINGS

The settings file is located in `./baobab/settings.py`. **Make sure you edit this file to setup your installation**.

In it, you can define some default data such user and admin accounts, as well as important options such as the `SECRET_KEY`, default time zones or to change `DEBUG=False` when not in development.

For production (actually, whenever `DEBUG is not True`) you need to authorize your hosts by editing `ALLOWED_HOSTS = []`.

More information on Django settings can be found [in the docs](https://docs.djangoproject.com/en/1.4/ref/settings).

#### Using a custom settings file

For example, create a Python package at `/etc/baobab/__init__.py` and your settings module file at `/etc/baobab/settings.py`.

You can make sure your package will be in your Python's path by exporting it

    $ export PYTHONPATH=/etc
    
Then you can use the `--settings` flag, like so:

    $ baobab <command> --settings baobab.settings
    
Alternatively, you can set Django's buit-in environment variable

    $ export DJANGO_SETTINGS_MODULE=baobab.settings
    $ baobab <command>

### BUILD THE WEB CLIENT

Install the web client's dependencies and build the app

    $ npm install
    $ npm run dev

### SETUP THE DATABASE

Load the schema and create the default data

    $ baobab setup-dev

By default, a SQLite database file will be created at `./baobab/default.db` if it doesn't exist.

### LAUNCH THE SERVER

To launch the web server with the admin interface and REST API :

    $ baobab runserver

Or  `$ ./baobab/bin/cmd_baobab.py runserver` if you've built the debian package.

Open your browser at `http://localhost:8000/` to see the website. The admin page is accessible at `http://localhost:8000/admin`.

### API USAGE

In addition to the web client, Baobab is also usable via a REST API that delivers content in JSON format. 

The API is self-described, so you can start exploring it by making a simple request to the main endpoint.

Example:

    $ curl http://localhost:8000/api | python -m json.tools

    {
        "events": {
            "list_endpoint": "/api/events",
            "schema": "/api/events/schema"
        },
        "services": {
            "list_endpoint": "/api/services",
            "schema": "/api/services/schema"
        },
        "status": {
            "list_endpoint": "/api/status",
            "schema": "/api/status/schema"
        }
    }


### UPDATING THE DATABASE MODEL

Follow this simple process to modify the database structure. Migrations are managed with [South](http://south.aeracode.org/).

1. Modify the application's models
2. Create schema migration files
3. Create data migration files if necessary
4. Apply the migrations

#### Changing the model

Update `models.py` in the appropriate app folder to add / remove fields and/or tables (for example `./baobab/backoffice/models.py`).

#### Creating a schema migration

    $ baobab schemamigration <namespace> --auto

**Tip**: Remember to use `db.rename_column` in a migration script to rename a field.

#### Creating a data migration 

    $ baobab datamigration <namespace> <data_migration_name>

#### Running migrations

    $ baobab migrate  

**Tip**: when switching branches, you might forget to rollback or apply relevant migrations. You can avoid that by adding a git hook. See some examples in`./misc/hooks/*`.

### OVERRIDING THE DEFAULT USER

You can override the default user login credentials by setting the `DEFAULT_USER_LOGIN` and `DEFAULT_USER_PASSWORD` variables.

If these variables are not set when the setup scripts are run, you will be prompted to create a default user.

### Social network integrations: Twitter, IRC, etc.

#### Custom integrations

At the moment baobab can publish status updates to Twitter and IRC, but you can easily add your own integrations.
You only need to create a new class in the `socialnetwork` app and inherit from the `SocialNetworkBase` class.

Each social network has its own configuration. Please take a look at the `settings.py` file for more information.

Upon the creation of an `Event`, a status update will be immediately published to each configured integration.

#### Maintenance

When the event is of the `Maintenance` type, baobab can publish a status update automatically at the estimated start date.
Simply create a cron task to execute `$ baobad social_network` to achieve this.

### TRANSLATIONS
 
You can easily translate content published with Baobab and retrieve translated content via the API.

Special user permissions can be granted for translators, and the translation interface is available in the backoffice at http://localhost:8000/admin/translate.`

You can then retrieve translated content by adding the "Accept-Language" header to your API requests. If you don't, Baobab automatically falls back to English (the default language).

For example:

    curl -H "Accept-Language: fr" http://localhost:8000/api/events

Please note that while the web client can consume translated content (browsers will automatically send the `Accept-Language` header according to the user's language), the web interface itself is not localised (nor localisable in the current state of the implementation).

### TESTING

You can run unit tests on specific namespaces or the whole app. No configuration is necessary for this.

    $ baobab test <namespace>

Note the `translate` namespace has no dedicated testing suite. The translation features are tested within the `apirest` namespace.

## PRODUCTION USAGE

In a production environment, you'll want to use different settings and technologies to better serve your app securily at scale.

We use Gunicorn as the app server, along with Nginx to proxy HTTP requests, instead of just running Django's webserver.

An example Gunicorn config for Baobab could look like this:

```
CONFIG = {
    'working_dir': '/srv/baobab',
    'environment': {
        'DJANGO_SETTINGS_MODULE': 'myown.settings',
        'PYTHONPATH': '/etc/'
    },
    'args': (
        '--bind=127.0.0.1:8008',
        '--workers=4',
        '--timeout=10',
        'baobab.wsgi:application',
    ),
}

```

Your production-ready settings file would then reside in `/etc/myown/settings.py`, along with the Python package file `/etc/myown/__init__.py`.

Also check the `nginx/sites-available/` folder for our sample Nginx configuration files.

You could also use MySQL or PostgreSQL as database backends instead of SQLite, for example.

### Manual deploy

After copying the application files to the server, install the web client's dependencies and build it for release (repeat whenever these assets change)

    $ npm install
    $ NODE_ENV=production npm run release

Then, setup the database and run migrations (repeat whenever the model changes)

    $ baobab syncdb
    
Now you can start or restart your chosen webserver solution.

### Debian package

Baobab can also be built into a Debian package, which is useful for production deployments on compatible systems. Gandi's own packaging details can be found in the `debian/` folder.

To build the package:

    $ debuild -us -uc -b || dpkg-buildpackage -us -uc -b

You can then place it on a Production server and install it. All scripts will be automatically launched. You can then start or restart the app and web servers.

## CONTRIBUTING

### Create issues

Any major changes should be documented as [a GitHub issue](#) before you start working on it.

### Proposing your changes

Don't hesitate -- we appreciate every contribution, no matter how small.

Create a git branch with your new feature or bugfix and either (in order of preference):

* open a Pull Request on GitHub
* mail the patch to feedback@gandi.net,
* send the URL for your branch and we will review/merge it if correct

We'll check your pull requests in the timeliest manner possible. If we can't accept your PR for some reason,
we'll give you feedback and you're encouraged to try again!

### Submission conventions

Fork the repository and make changes on your fork in a feature branch:

- If it's a bug fix branch, name it XXXX-something where XXXX is the number of the issue.
- If it's a feature branch, create an enhancement issue to announce your intentions, and name it XXXX-something where XXXX is the number of the issue.

#### Tests

Submit unit tests for your changes. Run the full test suite on your branch before submitting a pull request.

#### Documentation

Update the documentation when creating or modifying features.

## Code status

[![Build Status](https://travis-ci.org/Gandi/baobab.svg?branch=master)](https://travis-ci.org/Gandi/baobab)
[![Coverage Status](https://coveralls.io/repos/Gandi/baobab/badge.svg?branch=master)](https://coveralls.io/r/Gandi/baobab?branch=master)

## LICENSE

Please see the `LICENSE` file.

CHANGELOG
Baobab  changelog
=================

1.4
---

 * Debian packaging fix

1.3
---
  * css: use a real default front
  * css: avoid too big container
  * css: make article's paragraph more readable
  * api: html rendre: force new-line in <pre> to be able to see long text
  * update default TLS configuration
  * Use modern SSL ciphers from mozilla

1.2-1
-----
  * fix dependencies
  * fix test for travis
  * fix test running with py.test

1.2
---
  * refactoring the social network
  * the msg_id has been moved to its own table to handle all Socialnetwork
  * add a IRC bot:
    need to configured the information about the bot and the servers/chan
    in the new var IRC in the settings files

1.1
-----
  * add missing js file for the django's admin
  * fix js dependence: html5shiv upgrade from 3.7.2 to 3.7.3
  * fixed on Debian package build

1.0.0
-----
 * open sourcing

0.4.4
-----
 * front:
   - Add eventlogs in the public event page
   - Some CSS and HTML enhancements:
     - link to rss and back to gandi
     - roboto font used instead of browser's default serif font
     - align start / end dates
     - use shorter dates when they're less than 18 hours from now
     - background to #fafafa instead of #ffffff
   - Standardization on dates
   - Add markdown documentation link
 - Fix static file between django and nginx
 * rss: WARNING need a migration [BACKOFFICE]
  - add a field 'last_update' in 'Event'
  - datamigration to correctly set 'last_update'
  - entry is update if a new eventlog is added or if the date_end is defined
  - a tag [STARTED], [UPDATE] or [FINISHED] is added on an entry
  - <description> contains more information about the event:
    - eventlogs
    - last update
    - date start
    - date_end
    - impacted services
  - the 'last_update' field is added into the backoffice and the REST API
 * api: update /events to also output the 'last_update' field

0.4.3
-----
  * api: new filter for /events:
    - date_end=null to filter on 'date_end IS NULL'
    - current=true or current=false to get the current (or not) event

0.4.2
-----
 * the data send by the API are translatable
  - create a new app: translate
  - create a group: translate - traductor need to be in that group
 * fix the estimate_date_end not always updated
 * create a new module 'cron' to simplify the creation of new cron
 * add a new cron: 'close_event'
 * api: route /events add filter 'isnull' on a date_end
 * api: correctly handle error for filter category and services
        handle case insensitive for filter category and services
 * front: don't show 'Maintenance' event on timeline before they start
 * add RSS and Atom support
 * fix date return by the api: in some case microsecond was set

0.4.1
-----
 * API new value: FOGGY
   use when an on going incident doesn't impact any services
   it's only available as a global status and not for a service

0.3.8
-----
 * api rest better output on text/html:
  - add 2 variables in the settings:
   - TITLE_SCHEMA
   - TITLE_RESULT
 * api rest handle 404:
  - add a variable in the settings:
   - HTTP_NOT_FOUND

0.3.3
-----
 * apirest better handling on Exception
  - add 2 variables in the settings:
    HTTP_NOT_IMPLEMENTED_ERROR
    HTTP_APPLICATION_ERROR

0.3.2
-----
 * create a Twitter lib
   - need to update the settings:
      TWITTER_CONSUMER_KEY, TWITTER_CONSUMER_SECRET,
      TWITTER_ACCESS_TOKEN, TWITTER_ACCESS_TOKEN_SECRET,
      URL_EVENT
   - need to create a app in the twitter's noc account:
       https://apps.twitter.com/ (the Callback URL is not used)
   - need to allow the account itself to use the app
   - add a new dependence: python-oauth2

0.2.11
------
 * New dependency: django-htmlmin (django-htmlmin < 0.7.0)
 * Update settings:
  - add the varaible: HTML_MINIFY = True
  - update the varaible: MIDDLEWARE_CLASSES with:
    'django.middleware.gzip.GZipMiddleware'
    'htmlmin.middleware.HtmlMinifyMiddleware'
  - add the variable USE_ETAGS

0.0.2
-----
 * [api]
     change those routes:
       status: now return the global status: sunny, cloudy, stormy
       services: add a filed status
     only allow get request
     redfine the date serializer to always return a UTC datetime in iso 8601
 * [backoffice] WARNING need a migration
                a user can set its own timezone
 * [settings]
     update a variable to auto set the user's timezone
       MIDDLEWARE_CLASSES = (
         'baobab.utils.set_current_timezone.TimezoneMiddleware'
       )

0.0.1
-----
 * initial release


baobab/settings.py
# -*- coding: utf-8 -*-
# Django settings for baobab project.
from os.path import abspath, dirname, isfile, exists
import sys

# to locally override the default conf
import getpass
import imp

DEBUG = True
TEMPLATE_DEBUG = DEBUG
TEST = 'test' in sys.argv
HERE = dirname(abspath(__file__))
ROOT_DIR = HERE
HTML_MINIFY = True

USE_ETAGS = True
INFINITE_CACHE = False
SERVE_GZIP = False

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

# XXX need the trailing /
URL_EVENT = 'http://a.url/timeline/events/'

# twitter info
# TWITTER_CONSUMER_KEY = ""
TWITTER_CONSUMER_SECRET = ""

TWITTER_ACCESS_TOKEN = ""
TWITTER_ACCESS_TOKEN_SECRET = ""

TWITTER_MAX_CHAR = 140
TWITTER_URL_SHORTENER = 22
# XXX -1 for the space between the text and the url
TWITTER_ALLOWED_CHAR = TWITTER_MAX_CHAR - TWITTER_URL_SHORTENER - 1
TWITTER_URL_API = 'https://api.twitter.com/1.1/'

# IRC info
IRC = [
    {
        'nick': 'a_nick',
        'username': 'a_user_name',
        'ircname': 'a_irc_name',
        #'password': "",
        # 'port': 6697,
        # 'ssl': True,
        'server': 'a.server.irc',
        'channels': [
            {'name': '#chan', },
            {'name': '#chan2', 'password': 'pwd_chan2'},
        ],
        'prefix': '-all-',
        'on_welcome': 'Hi everyone',
        'on_quit': 'Going down for maintance will be back soon',
    }
]

IRC_SOCKET_PATH = '/run/baobab/baobab_irc.sock'

# see RFC http://tools.ietf.org/html/rfc2812#page-6
# -2 for the mandatory end '\r\n'
IRC_MAX_CHAR = 512 - 2
# get the longer prefix
LEN = max(map(lambda x: len(x.get('prefix', '')), IRC))
IRC_PREFIX_LEN = LEN + 1 if LEN else 0
# -10 for the id of the event
IRC_URL_LENGTH = len(URL_EVENT) - 10
# XXX -1 for the space between the text and the url
IRC_ALLOWED_CHAR = IRC_MAX_CHAR - IRC_PREFIX_LEN - IRC_URL_LENGTH - 1

# message used by the api
HTTP_NOT_IMPLEMENTED_ERROR = 'Sorry but it\'s not implemented yet\n'  # 501
HTTP_APPLICATION_ERROR = 'Sorry but something went wrong\n'  # 500
HTTP_NOT_FOUND = 'Resource not found\n'  # 404

# API HTML OUTPUT
TITLE_SCHEMA = 'Api schema'
TITLE_RESULT = 'Api result'

if DEBUG:
    # we are in a dev env, you can set a default user
    DEFAULT_USER_LOGIN = 'gandi'
    DEFAULT_USER_PWD = 'gandi'
    DEFAULT_USER_MAIL = 'devnull@gandi.net'

    # PROXY_HOST = ''
    # PROXY_PORT = 8080

MANAGERS = ADMINS

# XXX to avoid error in the prod env with the baoab command you need to localy
#     overwrite the DATABASES in conf/user_name.py
DATABASES = {}

if 'test' in sys.argv:
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
        }
    }

FIXTURE_DIRS = (
    '%s/fixtures/' % HERE,
)

# Hosts/domain names that are valid for this site; required if DEBUG is False
# See https://docs.djangoproject.com/en/1.4/ref/settings/#allowed-hosts
ALLOWED_HOSTS = []

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'Europe/Paris'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/home/media/media.lawrence.com/media/"
MEDIA_ROOT = '%s/static/' % HERE

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://media.lawrence.com/media/", "http://example.com/media/"
MEDIA_URL = '/static/'

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/home/media/media.lawrence.com/static/"
STATIC_ROOT = ''

# NEEDED FOR THE ADMIN
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    #    'django.contrib.staticfiles.finders.FileSystemFinder',
    #    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    #    'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = 'a secret key'

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    #     'django.template.loaders.eggs.Loader',
)

# 26/03/2015 - remove CommonMiddleware. This fix the issue of no content when
# serving static files.
MIDDLEWARE_CLASSES = (
    'django.middleware.gzip.GZipMiddleware',
    'baobab.middleware.CorsMiddleware',
    #'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'baobab.utils.set_current_timezone.TimezoneMiddleware'
)

ROOT_URLCONF = 'baobab.urls'

# Python dotted path to the WSGI application used by Django's runserver.
WSGI_APPLICATION = 'baobab.wsgi.application'

TEMPLATE_DIRS = (
    ''
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.admin',
    'south',
    # XXX need to be first to be able to path correctly for test
    #     See TestSocialNetwork.test test_10_get_max_char_ko
    'baobab.socialnetwork',
    'baobab.backoffice',
    'baobab.apirest',
    'baobab.front',
    'baobab.translate',
    'baobab.rss',
    'baobab.cron',
)

APPEND_SLASH = False
TASTYPIE_ALLOW_MISSING_SLASH = True
TASTYPIE_DEFAULT_FORMATS = ['json', 'html']
SOUTH_TESTS_MIGRATE = False
API_LIMIT_PER_PAGE = 0

if DEBUG or TEST:
    import warnings
    warnings.filterwarnings(
        'error', r'DateTimeField .* received a naive datetime',
        RuntimeWarning, r'django\.db\.models\.fields')

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'syslog': {
            'format': '%(name)s: [%(process)d] [%(levelname)s] %(message)s',
        },
        'console': {
            'format': '%(asctime)s [%(levelname)s]: %(name)s - %(message)s',
        }
    },
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        },
        'syslog': {
            'level': 'INFO',
            'class': 'logging.handlers.SysLogHandler',
            'address': '/dev/log',
            'formatter': 'syslog',
        },
        'console': {
            'class': 'logging.StreamHandler',
            'level': 'INFO',
            'stream': 'ext://sys.stdout',
            'formatter': 'console',
        }
    },
    'root': {
        'level': 'INFO',
        'handlers': [
            'console',
            'syslog',
        ],
        'propagate': True,
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
    }
}

if not exists('/dev/log'):
    # disable syslog for travis
    del LOGGING['formatters']['syslog']
    del LOGGING['handlers']['syslog']
    LOGGING['root']['handlers'].remove('syslog')

import mimetypes
mimetypes.add_type("application/vnd.ms-fontobject", ".eot", True)
mimetypes.add_type("application/font-woff", ".woff", True)
mimetypes.add_type("font/ttf", ".ttf", True)
mimetypes.add_type("font/otf", ".otf", True)


# allow override the default conf
conf_file = '%s/conf/%s.py' % (HERE, getpass.getuser())
if not isfile(conf_file) and DEBUG:
    conf_file = '%s/conf/default.py' % HERE

if isfile(conf_file):
    # the purpose here is to 'replace' the default 'baobab.settings' by
    # an other one but still keep it as 'baobab.settings' and
    # not a 'baobab.settings.conf.user_name'

    imp.load_source('baobab.settings', conf_file)
    # for python >= 3.3 do this
    # importlib.machinery.SourceFileLoader('baobab.settings',
    #                                      conf_file).load_module()
